x-pack/platform/plugins/shared/streams/server/lib/streams/state_management/types.ts

```diff
@@ -0,0 +1,58 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { IScopedClusterClient, Logger } from '@kbn/core/server';
+import { StreamDefinition, WiredStreamUpsertRequest } from '@kbn/streams-schema';
+import { State } from './state'; // Does this create a circular dependency?
+import { StreamsStorageClient } from '../service';
+
+export interface ValidationResult {
+  isValid: boolean;
+  errors: string[]; // Or Errors?
+}
+
+export type StreamChangeStatus = 'unchanged' | 'upserted' | 'deleted';
+export type StreamCommitStatus = 'uncomitted' | 'committing' | 'committed';
+
+// Interface or abstract class to make somethings private?
+// Should all of these methods be async from the start?
+export interface StreamActiveRecord {
+  definition: StreamDefinition;
+  changeStatus: StreamChangeStatus;
+  commitStatus: StreamCommitStatus;
+  clone(): StreamActiveRecord;
+  markForDeletion(desiredState: State, startingState: State): void;
+  update(newDefinition: StreamDefinition, desiredState: State, startingState: State): void;
+  validate(
+    desiredState: State,
+    startingState: State,
+    scopedClusterClient: IScopedClusterClient
+  ): Promise<ValidationResult>;
```

Comment:
One thing I'm a bit uncertain about is if the apply changes phase should reject changes or apply invalid state and then let the validation step catch it.